core/src/main/java/org/infinispan/marshall/exts/ArrayExternalizer.java

```diff
@@ -0,0 +1,114 @@
+package org.infinispan.marshall.exts;
+
+import java.io.IOException;
+import java.io.ObjectInput;
+import java.io.ObjectOutput;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.io.UnsignedNumeric;
+import org.infinispan.commons.marshall.AbstractExternalizer;
+import org.infinispan.commons.marshall.AdvancedExternalizer;
+import org.infinispan.commons.marshall.MarshallUtil;
+import org.infinispan.commons.util.Util;
+import org.infinispan.marshall.core.Ids;
+import org.jboss.marshalling.util.IdentityIntMap;
+
+/**
+ * Externalizers for diverse array types.
+ *
+ * @author Galder Zamarreño
+ * @since 6.0
+ */
+public class ArrayExternalizer extends AbstractExternalizer<Object> {
```

Comment:
There are times when we just don't know the type, e.g. `StreamResponseCommand.response`. This can be anything from a primitive to an array of entries, so we need some fallback for when type might be an array and we can't know for sure in advance. If we don't do anything, it'd fallback on the external marshaller.

Also, a user could store an array of a type for which there's an Externalizer. Without this support, the user would need to give us an externalizer for marshalling the array type, or it would be delegated to the external marshaller (JBoss Marshaller in this case).